sympy/sets/fancysets.py

```diff
@@ -307,41 +311,71 @@ def __new__(cls, flambda, *sets):
         if not isinstance(flambda, Lambda):
             raise ValueError('First argument must be a Lambda')
 
-        sets = [_sympify(s) for s in sets]
+        signature = flambda.signature
 
-        if flambda is S.IdentityFunction:
-            if len(sets) != 1:
-                raise ValueError('Identity function requires a single set')
-            return sets[0]
+        if len(signature) != len(sets):
+            raise ValueError('Incompatible signature')
+
+        sets = [_sympify(s) for s in sets]
 
         if not all(isinstance(s, Set) for s in sets):
             raise TypeError("Set arguments to ImageSet should of type Set")
 
-        sets = [s.flatten() if s.is_ProductSet else s for s in sets]
+        if not all(cls._check_sig(sg, st) for sg, st in zip(signature, sets)):
+            raise ValueError("Signature %s does not match sets %s" % (signature, sets))
+
+        if flambda is S.IdentityFunction and len(sets) == 1:
+            return sets[0]
 
         if not set(flambda.variables) & flambda.expr.free_symbols:
-            emptyprod = fuzzy_or(s.is_empty for s in sets)
-            if emptyprod == True:
+            is_empty = fuzzy_or(s.is_empty for s in sets)
+            if is_empty == True:
                 return S.EmptySet
-            elif emptyprod == False:
+            elif is_empty == False:
                 return FiniteSet(flambda.expr)
 
         return Basic.__new__(cls, flambda, *sets)
 
     lamda = property(lambda self: self.args[0])
+    base_sets = property(lambda self: self.args[1:])
 
     @property
     def base_set(self):
-        sets = self.args[1:]
+        # XXX: Maybe deprecate this? It is poorly defined in handling
+        # the multivariate case...
+        sets = self.base_sets
         if len(sets) == 1:
             return sets[0]
         else:
-            return ProductSet(*self.args[1:]).flatten()
+            return ProductSet(*sets).flatten()
+
+    @property
+    def base_pset(self):
+        return ProductSet(*self.base_sets)
+
+    @classmethod
+    def _check_sig(cls, sig_i, set_i):
```

Comment:
Is it possible to rewrite this function to be strictly iterative in nature. We are already hitting maximum recursion depth in a number of test in sympy?

Reply:
This can be made iterative but the recursion depth is typically 1 or 2. I don't see any reason why deep recursion would be needed here.